libdrgn/arch_s390.c

```diff
@@ -0,0 +1,327 @@
+	if (err)
+		return err;
+
+	err = drgn_type_find_member(psw_type.type, "addr", &member, addr_offset);
+	if (err)
+		return err;
+
+	return drgn_type_find_member(psw_type.type, "mask", &member, mask_offset);
+}
+
+/*
+ * s390 puts a stackframe + stack pt_regs on the kernel stack when entering
+ * program check handler, irq handlers or the system call handler.
+ */
+static struct drgn_error *
+unwind_check_pt_regs(struct drgn_program *prog,
+		     struct drgn_register_state *regs,
+		     struct drgn_register_state **ret)
+{
+	struct optional_uint64 tmp = drgn_register_state_get_u64(prog, regs, r15);
+	uint64_t psw_offset, mask_offset, addr_offset, gprs_offset, bc_offset;
+	uint64_t nextbc, frame_size, ptregs_size, *pswa, *pswm;
+	struct drgn_qualified_type frame_type;
+	struct drgn_type_member *member;
+	struct drgn_error *err;
+
+	if (!tmp.has_value || !tmp.value)
+		return &drgn_stop;
+
+	err = unwind_get_pt_regs_sizes(prog, &gprs_offset, &psw_offset, &addr_offset,
+				       &mask_offset, &frame_size, &ptregs_size);
+	if (err)
+		return err;
+
+	uint8_t *frame = malloc(frame_size + ptregs_size);
+	if (!frame)
+		return &drgn_enomem;
+
+	err = drgn_program_read_memory(prog, frame, tmp.value,
+				       frame_size + ptregs_size, false);
+	if (err)
+		goto out;
+
+	struct drgn_register_state *regs2 = drgn_register_state_create(bc, false);
+	if (!regs2)
+		goto out;
+
+	uint8_t *ptregs = frame + frame_size;
+	pswa = (uint64_t *)(ptregs + (psw_offset + addr_offset) / 8);
+	pswm = (uint64_t *)(ptregs + (psw_offset + mask_offset) / 8);
+
+        uint64_t *gprs = (uint64_t *)(ptregs + gprs_offset / 8);
+	drgn_register_state_set_range_from_buffer(regs2, r0, r15, gprs);
+	drgn_register_state_set_from_buffer(regs2, pswa, pswa);
+	drgn_register_state_set_from_buffer(regs2, pswm, pswm);
+	drgn_register_state_set_pc_from_register(prog, regs2, pswa);
+	drgn_register_state_set_from_buffer(regs2, bc, &gprs[15]);
+
+	*ret = regs2;
+
+	if (*pswm & (1UL << 48)) {
+		/* don't follow to userspace */
+	        return &drgn_stop;
+	}
+	return NULL;
+out:
+	free(frame);
+	return &drgn_enomem;
+}
+
+static struct drgn_error *
+fallback_unwind_s390(struct drgn_program *prog,
+		     struct drgn_register_state *regs,
+		     struct drgn_register_state **ret)
+{
+	struct optional_uint64 bc = drgn_register_state_get_u64(prog, regs, bc);
+	struct drgn_qualified_type frame_type;
+	struct drgn_type_member *member;
+	struct drgn_error *err;
+	uint64_t bit_offset;
+	uint8_t frame[256];
+
+	if (!bc.value)
+		return unwind_check_pt_regs(prog, regs, ret);
+
+	err = drgn_program_read_memory(prog, frame, bc.value, sizeof(frame), false);
+	if (err)
+		return err;
+
+	err = drgn_program_find_type(prog, "struct stack_frame", NULL, &frame_type);
+	if (err)
+		return err;
+
+	err = drgn_type_find_member(frame_type.type, "gprs", &member, &bit_offset);
+	if (err)
+		return err;
+
+	drgn_register_state_set_range_from_buffer(*ret, r6, r15, frame + bit_offset / 8);
```

Comment:
From my limited understanding based on reading the ABI documentation and looking at some disassembly, the only registers that are guaranteed to be saved in the frame are `r14` and `r15`; the rest may or may not be saved (other than for the initial frame saved in `__switch_to()`, which always saves `r6`-`r15`). Is that correct?

If so, this could lead us to report bogus register values to the user. Let's be conservative and only recover `r14` and `r15`.